### ISSUE TITLE
Setting from the CentralConfiguration (ie via Kibana) not visible via Agent.Config

### ISSUE THREAD

**Issue by pbalwall** [elastic/apm-agent-dotnet#1144]
Is there a way to view the latest APM Agent configuration, after incorporating the values retreived via the CentralConfiguration aspect of Agent configuration?

I am using Agent.Config, hoping that it would contain these config values, but it appears to solely have the configuration present in my web.config. I am using Agent 1.7.1.

Based on some debug logging done by the Agent, I can see that it does talk to the APMServer and does retrieve the current 'central configuration'. See below for a example message

Another question: from some of these messages (and from some docs) it implied that CaptureBodyContentTypes is also dynamically configurable (ie via central config). But the Kibana app does not allow its configuration?

--------------
{CentralConfigResponseParser} Processing HTTP response...
+-> Response:
    StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.StreamContent, Headers:
    {
      Access-Control-Expose-Headers: Etag
      X-Content-Type-Options: nosniff
      Cache-Control: must-revalidate, max-age=30
      Date: Mon, 25 Jan 2021 22:45:45 GMT
      ETag: "a087b409681d7542dbded4f4516c67664a3d8aed"
      Content-Length: 212
      Content-Type: text/plain; charset=utf-8
    }
+-> Response body [length: 212]:
    {"capture_body":"errors","capture_headers":"false","log_level":"info","recording":"false","span_frames_min_duration":"10NaNms","stack_trace_limit":"0","transaction_max_spans":"10","transaction_sample_rate":".5"}

**Comment by pbalwall** [comment]
To add more detail. It looks like the dotnet agent is indeed honoring the current central configuration. For example, if I set Recording=false in the central configuration (via Kibana), events dont get sent to Elasticsearch (as expected). My main issue is that the Agent.Config does not reflect the actual active configuration - it appears to be the static configuration present in web.config.

Is there a way to have access to the active configuration that the agent is acting upon?

**Comment by gregkalapos** [comment]
> Is there a way to view the latest APM Agent configuration, after incorporating the values retreived via the CentralConfiguration aspect of Agent configuration?

Not at this point. Most of these things are internal - each time a transaction is started the agent takes a snapshot of the config which is a merge of the current static config and the configs from the Kibana (if central config is active). It's [here](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Model/Transaction.cs#L207) in the code - but like I said, it's internal.

> Another question: from some of these messages (and from some docs) it implied that CaptureBodyContentTypes is also dynamically configurable (ie via central config). But the Kibana app does not allow its configuration?

Oh, that's not good. You are right, in the docs we state it's central configurable and indeed, it's not part of the settings in Kibana. I opened https://github.com/elastic/apm-agent-dotnet/issues/1146 to implement this.


So, yes, the agent honors those settings, but it does not provide a public API to read those. What's your exact use case? To be honest we haven't really thought about opening up this API, but we could consider it.

**Comment by pbalwall** [comment]
Hello Greg. 
The use case is really primarily to take advantage of the dynamic configurability, and based on config settings modulate the amount of instrumentation and diagnostic info included in the Custom items within a APM transaction. In our case we have a standard amount of APM instrumentation (ie the amount of custom metrics, etc.) that we place in the Labels items. But at times, especially if we know that a customer is having an performance issue, that we might want to add additional diagnostic/metrics info to help understand the issue - but because of the processing/network/storage aspect we do not want to do this at all times for everyone. In this case, I am hoping to use the LogLevel (or perhaps CaptureBody) and the CaptureBodyContentTypes (which thankfully is just a set of strings .. and I know this is getting hacky) to detect whether to turn on increased instrumentation and even to do so only for a particular customer.

So basically the use case is to increase the amount of custom instrumentation added into APM transactions (ie what is sent to elasticsearch) dynamically.

**Comment by pbalwall** [comment]
Sorry, forgot to add. In terms of needing a special API .. am wondering .. why wouldnt Agent.Config (which has access already to all the config params) not reflect the current operative configuration (ie the configuration values that APM internally is using)? Why would a client (ie my application) want solely what was originally present in the web.config file?  Ie if you are giving access to this configuration (via Agent.Config) why not present the actual current operative config?

**Comment by pbalwall** [comment]
@gregkalapos , just wondering if you saw my response on use case.

**Comment by glucaci** [comment]
Hi @gregkalapos ,

We are also taking look how we can fetch the current `LogLevel` to enable / disable other log components than the APM.

Could we make the access to the `ConfigStore` on the `AgentComponents` public and put it on `IApmAgent`? I would do a PR if it's ok.

Thanks!

**Comment by glucaci** [comment]
I've created a proposal for this https://github.com/elastic/apm-agent-dotnet/pull/1230

**Comment by gregkalapos** [comment]
Done in https://github.com/elastic/apm-agent-dotnet/pull/1323
